stdlib/LinearAlgebra/src/dense.jl

```diff
@@ -326,19 +326,32 @@ julia> kron(A, B)
  3+0im  0-3im  4+0im  0-4im
 ```
 """
-function kron(a::AbstractMatrix{T}, b::AbstractMatrix{S}) where {T,S}
-    require_one_based_indexing(a, b)
-    R = Matrix{promote_op(*,T,S)}(undef, size(a,1)*size(b,1), size(a,2)*size(b,2))
+function kron(A::AbstractMatrix{T}, B::AbstractMatrix{S}) where {T,S}
+    R = Matrix{promote_op(*,T,S)}(undef, size(A,1)*size(B,1), size(A,2)*size(B,2))
+    return kron!(R, A, B)
+end
+
+"""
+    kron!(C, A, B)
+
+Like [`kron`](@ref), but stores the results in `C`.
+"""
+function kron!(R::AbstractMatrix, A::AbstractMatrix, B::AbstractMatrix)
+    require_one_based_indexing(A, B)
     m = 0
-    @inbounds for j = 1:size(a,2), l = 1:size(b,2), i = 1:size(a,1)
-        aij = a[i,j]
-        for k = 1:size(b,1)
-            R[m += 1] = aij*b[k,l]
+    @inbounds for j = 1:size(A,2), l = 1:size(B,2), i = 1:size(A,1)
```

Comment:
You need to verify the size of `R`, especially as you are using `@inbounds` here. But also several times again below where there is no `@inbounds`, checking the size first sounds like a good idea; otherwise, providing too large an output matrix would go entirely unnoticed.

Reply:
Yeah, I didn't check the shape if match. I added them with `@boundscheck` if this make sense, so the original `kron` is not effected by this PR (this bounds checking is disabled in `kron` and can be disabled).

Reply:
I'm not really familiar with the bounds checking machinery, but this looks plausible to me.